assert: use same value equal for deepStrictEqual NaN

BridgeAR opened the pull request:
I think it makes sense to further improve assert a bit and to accept NaN as well. 

I also fixed the documentation a bit by adding missing changelogs to the assert docs.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
assert, doc

Trott commented on doc/api/assert.md:
```diff
@@ -121,9 +124,10 @@ changes:
-  comparison. (Which means they are free of the [caveats][]).
+1. Primitive values besides NaN are compared using the [Strict Equality
+   Comparison][] ( `===` ). Set and Map values, Map keys and NaN are compared
+   using the [SameValueZero][] comparison (Which means they are free of the
```
Nit: `Which` -> `which`

BridgeAR replied:
Addressed

vsemozhetbyt commented:
Maybe [this line](https://github.com/BridgeAR/node/blame/06003949a77c240b3390764b11f1276a96a67f24/doc/api/assert.md#L164) with exposed markdown syntax in JS example worth fixing? It can be confusing a bit.

BridgeAR commented:
@vsemozhetbyt thx, fixed

Trott commented on test/parallel/test-assert-deep.js:
```diff
@@ -466,4 +466,10 @@ assertOnlyDeepEqual(
 
+// Handle NaN
+{
+  assert.throws(() => assert.deepEqual(NaN, NaN), assert.AssertionError);
```
Nit: Leaving out the braces around the arrow function adds an implicit return. Since the returned value is not used, I prefer to add braces. Feel free to ignore, but my two cents on it.

Trott commented on test/parallel/test-assert-deep.js:
```diff
@@ -466,4 +466,10 @@ assertOnlyDeepEqual(
+// Handle NaN
+{
+  assert.throws(() => assert.deepEqual(NaN, NaN), assert.AssertionError);
+  assert.doesNotThrow(() => assert.deepStrictEqual(NaN, NaN));
```
Since tests are a form of documentation, maybe add an additional test case or two that demonstrates the utility a bit more:

```js
assert.doesNotThrow(() => { assert.deepStrictEqual({ a: NaN }, { a: NaN });
assert.doesNotThrow(() => { assert.deepStrictEqual( [ 1, 2, NaN, 4 ], [ 1, 2, NaN, 4 ]);
```

BridgeAR commented:
I updated the tests according to your suggestions @Trott

vsemozhetbyt commented on doc/api/assert.md:
```diff
@@ -102,6 +102,9 @@ parameter is undefined, a default error message is assigned.
 added: v1.2.0
 changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/12142
```
Is this the proper link?

BridgeAR replied:
Good catch :smiley:

refack commented on test/parallel/test-assert-deep.js:
```diff
@@ -466,4 +466,11 @@ assertOnlyDeepEqual(
 }
 
+// Handle NaN
+assert.throws(() => { assert.deepEqual(NaN, NaN); }, assert.AssertionError);
```
This is not intuitive. I'd assume that
![image](https://user-images.githubusercontent.com/96947/29738113-c66fbb3a-89e8-11e7-9402-c2e91e282abe.png)
(deepStrictEqual should be more strict than deepStrictEqual).
What's the rationale for not allowing `assert.deepEqual(NaN, NaN)`?

BridgeAR replied:
That is the bad part about the wording with `strict`. The strict stands for more exact in this case. `deepEqual` is actually more like a `looksSimilar` and it actually fails at seeing the similarity in this case even though it exists. I can add that check to deepEqual as well but I think it would be weird to have it there (even though I would call deepEqual broken anyway - we actually also use a own lint rule to prevent usage of it!).

refack replied:
Actually that is IMHO a great idea, create a better named superset assertion. That would also make this change semver-minor.

`sameValues` might be a better name (that way neither the word "strict" nor "deep" are necessary)

BridgeAR replied:
I think this is absolutely not the right thing to do. And as I tried to explain earlier `deepStrictEqual` is not strictly a superset of `deepEqual` because it was just a poor naming from the beginning on.  Now we have to live with that naming. Most people do for example not know that `NaN` is not equal to itself when compared with `strict equality`. And they also do not know about the other existing equality comparisons in JS. They mostly do not even know how `abstract equality` is defined - at least that is what my experience tells me.
The comparisons in general are well described [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness). Adding a new API would not help anyone and would instead create more confusion.

refack replied:
I agree that `assert.deepStrictEqual(NaN,NaN)` makes sense intuitively. 
But ECMAScript `abstract equal` is "broken" in this sense (`(NaN == NaN) === false`) and `strict equal` is broken (`(NaN === NaN) === false`) is the exact same way.
So IMHO we should be consistent with the standard if we keep using the term `equal` in our end-point names.

That's why I think breaking away for legacy with a `same` endpoint might be a good idea.

BridgeAR replied:
The deepEqual and deepStrictEqual functions have much more then just the "basic" comparison and therefore do not correspond to these properly. I would love if that would be the only distinction because in that case `deepEqual` would not be so badly broken but that is not the case and it will not change either.
Adding another random name makes the issue not less an issue but just added more problems.

lpinca commented on lib/assert.js:
```diff
@@ -167,8 +167,11 @@ function isObjectOrArrayTag(tag) {
-  if (actual === null || expected === null ||
-    typeof actual !== 'object' || typeof expected !== 'object') {
+  if (typeof actual !== 'object') {
+    return typeof actual === 'number' && Number.isNaN(actual) &&
```
I think `typeof actual === 'number'` can be removed, `Number.isNaN(actual)` is sufficient.

BridgeAR replied:
That is correct but it is faster for all other types to check for number first and if it is a number it is not a real penalty because that check is super fast.

lpinca replied:
Does it makes sense to do the same for `expected` then?

BridgeAR replied:
We could but I feel like that is overdoing it as it would only apply for the case where `Number.isNaN(actual)`, especially as I would expect `expected` to also be NaN in that case more often then not being of type number.

BridgeAR commented:
This will actually also fix a issue in a N-API test case that expects NaN to be equal to each other and they tested it by using deepStrictEqual.
```
AssertionError [ERR_ASSERTION]: [Number: NaN] deepStrictEqual [Number: NaN]
    at Object.<anonymous> (/home/ruben/repos/node/node/test/addons-napi/test_conversions/test.js:119:8)
```
I personally think this could be counted as a bug fix instead of a semver-major because I highly doubt that anyone relies on `NaN !== NaN` in a `deepStrictEqual` call. Especially as most people are not even aware that that is the case.
@jasnell @Trott what do you think about that?

Trott commented:
> I highly doubt that anyone relies on 

Changing this behavior *could* cause tests to succeed erroneously where two calculations are expected to come up with the same number but a bug in the code causes them both to come up with `NaN` instead.  The test may even include a guard using `typeof` but that will *still* turn up `number`.

I'd prefer it be treated as `semver-major` out of caution but it's a mild preference. If others feel more strongly that it shouldn't be treated as a breaking change, I'm OK with that.

refack commented:
IMHO it's a bug fix of the ad hoc spec (I see the current behavior of `assert` as a self defining spec) so I tend towards 'semver-major'.
On the other hand @BridgeAR pointed out to me that other 3rd party asset libraries already treat NaN as equal to itself.

tniessen commented on doc/api/assert.md:
```diff
@@ -121,9 +124,10 @@ changes:
-  ( `===` ). Set values and Map keys are compared using the [SameValueZero][]
-  comparison. (Which means they are free of the [caveats][]).
+1. Primitive values besides NaN are compared using the [Strict Equality
+   Comparison][] ( `===` ). Set and Map values, Map keys and NaN are compared
```
We usually use monospacing for constants such as `NaN` in the docs.

BridgeAR replied:
Addressed

tniessen commented:
>  But ECMAScript abstract equal is "broken" in this sense ((NaN == NaN) === false) and strict equal is broken ((NaN === NaN) === false) is the exact same way.

@refack No, it's not. I actually considered implementing `assert` support for `NaN` a couple of weeks ago, but came across this part of the famous IEEE-754 specs:

> Four mutually exclusive relations are possible: less than, equal, greater than, and unordered. The last arises when at least one operand is NaN. Every NaN shall compare unordered with everything, including itself. Comparisons shall ignore the sign of zero (so +0 = −0). Infinite operands of the same sign shall compare equal.

refack commented:
> @refack No, it's not. I actually considered implementing assert support for NaN a couple of weeks ago, but came across this part of the famous IEEE-754 specs:
> 
> Four mutually exclusive relations are possible: less than, equal, greater than, and unordered. The last arises when at least one operand is NaN. Every NaN shall compare unordered with everything, including itself. Comparisons shall ignore the sign of zero (so +0 = −0). Infinite operands of the same sign shall compare equal.

@BridgeAR pointed me to that spec as well. IMHO that spec should not have been implemented verbatim in JS, but that's done (BTW: If I read this correctly then `(NaN != NaN)` should be `false` as well)

My point was that if `(NaN == NaN) ≡ (NaN === NaN)` then we should make `deepEqual(NaN, NaN) ≡ deepStrictEqual(NaN, NaN)`, but @BridgeAR convinced me that `deepEqual` is not worth our hassle.

BridgeAR commented:
@Trott 
> Changing this behavior could cause tests to succeed erroneously where two calculations are expected to come up with the same number but a bug in the code causes them both to come up with NaN instead.

But in that case the test should not have existed _yet_ as it would have otherwise failed and the bug got hopefully fixed? So it is only something for the future, right?

BridgeAR commented:
I rebased due to conflicts and addressed the monospacing.

BridgeAR commented:
CITGM https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/960/

Trott commented:
> But in that case the test should not have existed yet as it would have otherwise failed and the bug got hopefully fixed? So it is only something for the future, right?

I'm thinking of regressions where a bug is introduced while refactoring or something like that. Might be unlikely as you'd need to mess up *two* calculations that way simultaneously, I suppose.

BridgeAR commented:
Rebased due to conflicts. I also fixed a documentation bug from a former commit where the PR number was wrong.

@Trott I am not sure I can follow. If I read your comment correct you would rather keep it as semver-major, right?

Trott commented:
>  I am not sure I can follow

@BridgeAR I'd prefer it be treated as semver-major out of caution but it's a very mild preference. If no one else agrees, I'm fine with my mild preference being disregarded. :-D

BridgeAR commented:
Landed in 1789dcfc873eea0b993a1fa20f2e30819a57b9c9 and ea2e6363f221c1c6ca5b04b295bd6d17ecca355b.
